File: c++/3000.cpp
// Link: https://leetcode.com/problems/number-of-zero-filled-subarrays/description/
// Runtime: 0 ms
// Memory: 29.28 MB

class Solution {
public:
  int areaOfMaxDiagonal(const std::vector<std::vector<int>> &dimensions) {
    int maxArea = 0, maxAlmostDiagonal = 0;
    for (int i = 0; i < static_cast<int>(dimensions.size()); i++) {
      int almostDiagonal = dimensions[i][0] * dimensions[i][0] +
                           dimensions[i][1] * dimensions[i][1];
      if (almostDiagonal > maxAlmostDiagonal) {
        maxAlmostDiagonal = almostDiagonal;
        maxArea = dimensions[i][0] * dimensions[i][1];
      } else if (fabs(almostDiagonal - maxAlmostDiagonal) <
                 std::numeric_limits<double>::epsilon()) {
        maxArea = std::max(maxArea, dimensions[i][0] * dimensions[i][1]);
      }
    }

    return maxArea;
  }
};

File: python/3000.py
# Link: https://leetcode.com/problems/maximum-area-of-longest-diagonal-rectangle/description/
# Runtime: 0 ms
# Memory: 17.78 MB


class Solution:
    def areaOfMaxDiagonal(self, dimensions: List[List[int]]) -> int:
        max_area, max_almost_diagonal = 0, 0
        for i, dimension in enumerate(dimensions):
            almost_diagonal = dimension[0] ** 2 + dimension[1] ** 2
            if almost_diagonal > max_almost_diagonal:
                max_almost_diagonal = almost_diagonal
                max_area = dimension[0] * dimension[1]
            elif almost_diagonal == max_almost_diagonal:
                max_area = max(max_area, dimension[0] * dimension[1])

        return max_area
